src/Database/PostgreSQL/Transact/Cluster.hs

```diff
@@ -91,19 +100,34 @@ class ExecutionMode modeConn modeQuery where
         m a
 
 
+runReadOnly ::
+    MonadBaseControl IO m =>
+    RunInBase m IO ->
+    CDBT 'ReadOnly m r ->
+    Connection ->
+    IO (StM m r)
+runReadOnly run (CDBT task) conn =
+    PSQL.withTransactionModeRetry txMode isSerializationAnomaly conn q
+  where
+    txMode = TransactionMode RepeatableRead PSQL.ReadOnly
+    isSerializationAnomaly SqlError{sqlErrorMsg} =
+        any (BS.isSuffixOf "concurrent update") $ BS.inits sqlErrorMsg
```

Comment:
Why do this inits thing? Can't you just do `isSuffixOf`?

Reply:
Ideally there would be a numeric error code for this problem.  However, it looks like we have to parse an error string.  This reduces to `isSuffixOf` when the error message exactly matches the docs.  If there is e.g. trailing whitespace or some other weirdness, this still finds the phrase.